### PR TITLE
Fix envelope of polygon containing pole.

### PR DESCRIPTION
**Pull request by awulkiew**
This PR fixes spherical and geographic envelope for polygons containing poles. Currently the envelope of a polygon is the envelope of the linear ring. If a polygon is a polar cap then the result is incorrect because the envelope should contain the pole even if it's above or below linear ring.

If an envelope covers the whole longitude range then poles are checked to see whether or not they are inside the ring. This check is done with `point_in_poly` strategy with custom `pole_side` strategy to avoid costly calculation done in regular side strategies.

Additional changes:
- envelope strategy previously used for linear and polygonal geometries as well as multi geometries is divided into three strategies: one for linestring, one for ring and one for range of boxes which is used for multi geometries,
- `point_iterator` and `segment_iterator` are no longer used (ranges are processed in strategies based on CS),
- `clockwise_view`, `closed_view` and `closed_clockwise_view` are adapted to geometry concepts based on the input `Range` which allows passing them as geometries into various algorithms,
- test are added and other that were previously disabled are enabled.

### CODE REVIEW

**Comment by barendgehrels** [comment]
Thanks! I'm not ready yet. My remarks so far are all minor of course.
It seems quite a change for such an issue. But you added a useful description to this PR, will continue later.

File: include/boost/geometry/strategy/spherical/envelope_range.hpp [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    {
+        geometry::set<1>(point, constants_t::min_latitude());
+    }
+    geometry::detail::closed_clockwise_view<Ring const> view(ring);
+    return point_in_range(point, view, pole_within_strategy) > 0;
+}
+
+template
+<
+    typename Range, typename Box,
```
Comment (vissarion) [inline include/boost/geometry/strategy/spherical/envelope_range.hpp:185]: According to our guidelines, long template parameter list shall be formatted with one template parameter per line.

File: include/boost/geometry/strategy/spherical/envelope_range.hpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+            break;
+        }
+    }
+
+    return strategy.result(state);
+}
+
+
+template <typename Ring, typename PoleWithinStrategy>
+inline bool pole_within(bool north_pole, Ring const& ring,
```
Comment (vissarion) [inline include/boost/geometry/strategy/spherical/envelope_range.hpp:154]: Alternatively, of having a `bool` parameter and an `if` statement you could directly pass `max_latitude` or `min_latitude` as a function parameter and explicitly set it in the function.

**Comment by vissarion** [comment]
Thanks for this PR! I left some comments.

File: include/boost/geometry/strategy/spherical/envelope_range.hpp [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        using point_t = typename boost::range_value<Range>::type;
+        geometry::model::referring_segment<point_t const> const seg(*prev, *it);
+        expand_strategy.apply(mbr, seg);
+        prev = it;
+        ++it;
+    }
+}
+
+
+// Side of pole WRT segment which doesn't contain it.
```
[the remark this answers is not in the document]
Comment (awulkiew) [reply include/boost/geometry/strategy/spherical/envelope_range.hpp:75]: Right, technically in de9im sense a segment of a polygon doesn't intersect a pole. In other words a polygon either contains the pole or they are disjoint.

File: include/boost/geometry/strategy/spherical/envelope_range.hpp [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+                             || math::equals(s_lon_diff, pi);
+        // Side of vertical segment is 0 for both poles.
+        if (s_vertical)
+        {
+            return 0;
+        }
+
+        // This strategy shouldn't be called in this case but just in case
+        // check if segment starts at a pole
+        if (math::equals(lat, lat1) || math::equals(lat, lat2))
```
[the remark this answers is not in the document]
Comment (awulkiew) [reply include/boost/geometry/strategy/spherical/envelope_range.hpp:115]: This should condition shouldn't be met aside from some unexpected case. The previous condition `s_vertical` however is perfectly normal. So I'd expect `s_vertical` to be true more often and therefore less conditions tested on average.

File: test/algorithms/envelope_expand/envelope_on_spheroid.cpp [hunk 5]
```diff
@@ -2493,16 +2493,18 @@ BOOST_AUTO_TEST_CASE( envelope_cw_ring )
...
     typedef test_envelope_on_sphere_or_spheroid<G, B> tester;
 
-    //double const eps = std::numeric_limits<double>::epsilon();
+    double const eps = std::numeric_limits<double>::epsilon();
 
     tester::apply("r01cw",
                   from_wkt<G>("POLYGON((0 10,0 45,50 10,0 10))"),
                   0, 10, 50, 45);
-#if 0
+
```
[the remark this answers is not in the document]
Comment (awulkiew) [reply test/algorithms/envelope_expand/envelope_on_spheroid.cpp:2501]: Some of the results may be correct by coincidence.

**Comment by awulkiew** [comment]
@vissarion Thanks for the review! I addressed your comments.

@barendgehrels What do you think about this PR?

**Comment by barendgehrels** [comment]
> @vissarion Thanks for the review! I addressed your comments.
> 
> @barendgehrels What do you think about this PR?

Yes, sorry, I forgot to finish the review. In general I'm fine with it (though I prefer separating fixes and refactoring - this makes it harder to judge).

I'll approve, but there are still some open questions from my side unanswered.

**Review by barendgehrels — APPROVED** [review]

File: include/boost/geometry/strategies/envelope/cartesian.hpp [hunk 6]
```diff
@@ -36,44 +37,54 @@ struct cartesian
...
 
     template <typename Geometry, typename Box>
     static auto envelope(Geometry const&, Box const&,
-                         typename util::enable_if_geometry_collection_t<Geometry> * = nullptr)
+                         std::enable_if_t
+                            <
+                                util::is_multi_linestring<Geometry>::value
+                             || util::is_multi_polygon<Geometry>::value
+                             || util::is_geometry_collection<Geometry>::value
+                            > * = nullptr)
```
[the remark this answers is not in the document]
Comment (awulkiew) [reply include/boost/geometry/strategies/envelope/cartesian.hpp:85]: This is not algorithm but envelope strategy getter in `cartesian` umbrella strategy.

File: include/boost/geometry/strategies/envelope/spherical.hpp [hunk 7]
```diff
@@ -13,13 +13,14 @@
 
 #include <type_traits>
 
-#include <boost/geometry/strategy/spherical/envelope.hpp>
+#include <boost/geometry/strategy/spherical/envelope.hpp> // Not used, for backward compatibility
```
[the remark this answers is not in the document]
Comment (awulkiew) [reply include/boost/geometry/strategies/envelope/spherical.hpp:16]: We probably don't. I'll remove it.

File: include/boost/geometry/strategy/cartesian/envelope_boxes.hpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+namespace strategy { namespace envelope
+{
+
+class cartesian_boxes
+{
+public:
+    template <typename Box>
+    class state
+    {
+        friend cartesian_boxes;
```
[the remark this answers is not in the document]
Comment (awulkiew) [reply include/boost/geometry/strategy/cartesian/envelope_boxes.hpp:28]: To allow the `cartesian_boxes` strategy to access the members of the `state` and at the same time prevent a caller from accessing them. This way we can be sure that the `state` is not going to be modified outside of this strategy.

File: include/boost/geometry/strategy/geographic/envelope_range.hpp [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        detail::spheroidal_linestring(range, mbr,
+                                      envelope::geographic_segment
+                                        <
+                                            FormulaPolicy, Spheroid, CalculationType
+                                        >(m_spheroid),
+                                      expand::geographic_segment
+                                        <
+                                            FormulaPolicy, Spheroid, CalculationType
+                                        >(m_spheroid));
```
[the remark this answers is not in the document]
Comment (awulkiew) [reply include/boost/geometry/strategy/geographic/envelope_range.hpp:56]: Would this be ok?
```
detail::spheroidal_linestring(range, mbr,
    envelope::geographic_segment
        <
            FormulaPolicy, Spheroid, CalculationType
        >(m_spheroid),
    expand::geographic_segment
        <
            FormulaPolicy, Spheroid, CalculationType
        >(m_spheroid));
```
or maybe this?
```
auto const envelope_s = envelope::geographic_segment
    <
        FormulaPolicy, Spheroid, CalculationType
    >(m_spheroid);
auto const expand_s = expand::geographic_segment
    <
        FormulaPolicy, Spheroid, CalculationType
    >(m_spheroid);
detail::spheroidal_linestring(range, mbr, envelope_s, expand_s);
```

[the remark this answers is not in the document]
Comment (barendgehrels) [reply include/boost/geometry/strategy/geographic/envelope_range.hpp:56]: Yes, that looks better. Both of them, but for me the last one is the most readable.
Thank you.

[the remark this answers is not in the document]
Comment (awulkiew) [reply include/boost/geometry/strategy/geographic/envelope_range.hpp]: Done.

**Review by vissarion — APPROVED** [review]
I am OK. Thanks!

**Pull request merged**